python/src/ops.cpp

```diff
@@ -3105,6 +3105,98 @@ void init_ops(py::module_& m) {
             array: The convolved array.
       )pbdoc");
   m.def(
+      "convNd",
+      [](const array& input,
+         const array& weight,
+         const std::variant<int, std::vector<int>>& stride,
+         const std::variant<int, std::vector<int>>& padding,
+         const std::variant<int, std::vector<int>>& kernel_dilation,
+         const std::variant<int, std::vector<int>>& input_dilation,
+         int groups,
+         bool flip,
+         StreamOrDevice s) {
+        std::vector<int> stride_vec;
+        std::vector<int> padding_vec;
+        std::vector<int> kernel_dilation_vec;
+        std::vector<int> input_dilation_vec;
+
+        if (auto pv = std::get_if<int>(&stride); pv) {
+          stride_vec.push_back(*pv);
+        } else {
+          stride_vec = std::get<std::vector<int>>(stride);
+        }
+
+        if (auto pv = std::get_if<int>(&padding); pv) {
+          padding_vec.push_back(*pv);
+        } else {
+          padding_vec = std::get<std::vector<int>>(padding);
+        }
+
+        if (auto pv = std::get_if<int>(&kernel_dilation); pv) {
+          kernel_dilation_vec.push_back(*pv);
+        } else {
+          kernel_dilation_vec = std::get<std::vector<int>>(kernel_dilation);
+        }
+
+        if (auto pv = std::get_if<int>(&input_dilation); pv) {
+          input_dilation_vec.push_back(*pv);
+        } else {
+          input_dilation_vec = std::get<std::vector<int>>(input_dilation);
+        }
+
+        return convNd(
+            /* const array& input = */ input,
+            /* const array& weight = */ weight,
+            /* std::vector<int> stride = */ stride_vec,
+            /* std::vector<int> padding = */ padding_vec,
+            /* std::vector<int> kernel_dilation = */ kernel_dilation_vec,
+            /* std::vector<int> input_dilation = */ input_dilation_vec,
+            /* int groups = */ groups,
+            /* bool flip = */ flip,
+            s);
+      },
+      "input"_a,
+      "weight"_a,
+      py::pos_only(),
+      "stride"_a = 1,
+      "padding"_a = 0,
+      "kernel_dilation"_a = 1,
+      "input_dilation"_a = 1,
+      "groups"_a = 1,
+      "flip"_a = false,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+        convNd(input: array, weight: array, /, stride: Union[int, List[int]] = 1, padding: Union[int, List[int]] = 0, kernel_dilation: Union[int, List[int]] = 1, input_dilation: Union[int, List[int]] = 1, groups: int = 1, flip: bool = false, *, stream: Union[None, Stream, Device] = None) -> array
+
+        ND convolution over an input with several channels
```

Comment:
Wow, didn't know you were adding an ND. How many dimensions does it go up to?

Reply:
I'm wondering about the decision to expose this if we only support 1 and 2d.. maybe it's useful, but I also wonder if it will just confuse people when they try to do 3d conv (since we have 1d and 2d in the API already, it seems like this should be used for >2d). 

I might suggest we put this aside in a branch (revive it from history) until we have 3d. But if you really prefer to keep it, I would at least put up a big note in the doc-string about only support for 1 and 2d

Reply:
The thinking was that we have a general conv function below - if you think the name `convNd` is confusing, I can switch it to something like `conv_general`

Reply:
I don't have a problem with the name. I am only worried that it's prematurely added. It's suggestive that it would work for any dimensions (e.g. 3) but it only works for 1 and 2.

I feel someone will see it and think they can do 3d convs in MLX but they can't yet. So I think either of these is fine:

1. put this aside in a branch (revive it from history) until we have 3d+.
2. Or put up a big note in the doc-string about only support for 1 and 2d

Though the first would probably be cleaner...but you can decide